ClientRuntimes/CSharp/Microsoft.Rest.ClientRuntime/BatchDelegatingHandler.cs

```diff
@@ -0,0 +1,185 @@
+    public class BatchDelegatingHandler : DelegatingHandler
+    {
+        /// <summary>
+        /// HTTP method (e.g. POST or PUT) to use in issuing a batch call.
+        /// </summary>
+        private readonly HttpMethod batchMethod;
+
+        /// <summary>
+        /// URL to issue the batch call to.
+        /// </summary>
+        private readonly Uri batchURL;
+
+        /// <summary>
+        /// List of operations to issue in a batch.
+        /// </summary>
+        private List<HttpRequestMessage> requests;
+
+        /// <summary>
+        /// Array of HTTP content returned by the server in response to operations that are issued in a batch.
+        /// </summary>
+        private HttpContent[] responses;
+
+        /// <summary>
+        /// Event that tells all tasks that they can proceed.
+        /// </summary>
+        private ManualResetEvent batchCompleted;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BatchDelegatingHandler"/> class.
+        /// </summary>
+        /// <param name="method">the type of method that the Batch operation is</param>
+        /// <param name="url">the URL that the Batch operation needs to hit</param>
+        public BatchDelegatingHandler(HttpMethod method, Uri url)
+        {
+            this.batchMethod = method;
+            this.batchURL = url;
+            this.Reset();
+        }
+
+        /// <summary>
+        /// Resets the batch handler to accept a new set of batch requests.
+        /// </summary>
+        public void Reset()
+        {
+            this.requests = new List<HttpRequestMessage>();
+            this.responses = null;
+            this.batchCompleted = new ManualResetEvent(false);
+        }
+
+        /// <summary>
+        /// Send all the queued up requests as a single batch request to the inner handler.
+        /// </summary>
+        /// <param name="cancellationToken">Cancellation token for the batch request</param>
+        /// <param name="boundary">Optional string which will be used as the encapsulation boundary between each request in multipart content; if not specified, a random GUID will be used</param>
+        /// <returns>Task for the batch</returns>
+        public async Task IssueBatch(CancellationToken cancellationToken = default(CancellationToken), string boundary = null)
+        {
+            // There must be at least one request to issue
+            if (this.requests.Count == 0)
+            {
+                throw new Exception("No queued requests to issue.");
+            }
+
+            // Sanity check. There must not be any queued responses from the server.
+            if (this.responses != null)
+            {
+                throw new Exception("This batch handler instance has been reused without calling reset first.");
+            }
+
+            // Create a request to the batch URL
+            HttpRequestMessage batchRequest = new HttpRequestMessage(this.batchMethod, this.batchURL);
```

Comment:
After discussing with @markcowl, the solution that he came up with would be to determine the batchURL based on the this.requests property (e.g. take the host from the URI of the first request in the collection and append "$batch"). This would accomplish two things:
1. The consumer doesn't need to worry about the batchURL, since it automatically comes from the ServiceClient
2. Make sure the consumer can't specify a batchURL with a host that is different from the host URIs of the individual requests

I don't think it's possible for the requests to have different hosts, simply because of how the requests are created (from the client), but just to be safe, it would be nice to validate that all the requests are being made to that same host.